R/target_market_share.R

```diff
@@ -282,105 +188,65 @@ target_market_share <- function(data,
+  data <- format_output_dataframe(data)
 
   corporate_economy <- calculate_ald_benchmark(ald, region_isos, by_company)
 
-  relevant_sectors <- unique(data$sector)
-
-  relevant_corporate_economy <- corporate_economy %>%
-    filter(.data$sector %in% relevant_sectors)
+  relevant_corporate_economy <- filter(
+    corporate_economy,
+    .data$sector %in% unique(data$sector)
+    )
 
   data %>%
     dplyr::bind_rows(relevant_corporate_economy) %>%
     ungroup()
 }
 
-pivot_wider2 <- function(data, ...) {
-  abort_if_has_list_colums(data)
-
-  out <- suppressWarnings(pivot_wider(data, ...))
-  unnest_list_columns(out)
-}
-
-unnest_list_columns <- function(data) {
-  if (utils::packageVersion("tidyr") < "1.1.2") {
-    suppressWarnings(unnest(data))
-  } else {
-    unnest(data, where(is.list))
-  }
-}
-
-tmsr_or_smsp <- function() {
-  dplyr::tribble(
-    ~which_metric, ~green_or_brown,
-    "tmsr_target_production", "brown",
-    "smsp_target_production", "green"
-  )
-}
+calculate_targets <- function(data){
+  target_groups <- c("sector_ald", "scenario", "region", "name_ald")
 
-separate_metric_from_name <- function(data) {
   data %>%
+    arrange(.data$year) %>%
+    group_by(!!!rlang::syms(c(target_groups, "technology", "year"))) %>%
+    mutate(technology_production = sum(.data$production)) %>%
+    group_by(!!!rlang::syms(c(target_groups, "year"))) %>%
+    mutate(sector_production = sum(.data$production)) %>%
+    group_by(!!!rlang::syms(c(target_groups, "technology"))) %>%
     mutate(
-      name = sub("weighted_", "", .data$name),
-      name = sub("(production)_", "\\1-", .data$name),
-      name = sub("(technology_share)_", "\\1-", .data$name)
+      initial_technology_production = first(.data$technology_production),
+      technology_production = NULL
     ) %>%
-    tidyr::separate(.data$name, into = c("name", "metric"), sep = "-")
-}
-
-maybe_add_name_ald <- function(data, by_company = FALSE) {
-  out <- data
-
-  if (by_company) {
-    out <- c(data, "name_ald")
-  }
-
-  return(out)
-}
-
-abort_if_has_list_colums <- function(data) {
-  if (has_list_colum(data)) {
-    abort("`data` must have no list column.")
-  }
-
-  invisible(data)
-}
-
-has_list_colum <- function(data) {
-  any(vapply(data, is.list, logical(1)))
+    group_by(!!!rlang::syms(target_groups)) %>%
+    mutate(
+      initial_sector_production = first(.data$sector_production),
+      sector_production = NULL) %>%
+    ungroup() %>%
+    mutate(
+      tmsr_target_production = .data$initial_technology_production *
+        .data$tmsr,
+      smsp_target_production = .data$initial_technology_production +
+        (.data$initial_sector_production * .data$smsp),
+      tmsr = NULL,
+      initial_technology_production = NULL,
+      initial_sector_production = NULL,
+      smsp = NULL
+    ) %>%
+    mutate(
+      smsp_target_production = ifelse(
+        .data$smsp_target_production < 0,
+        0,
+        .data$smsp_target_production
+      )
+    ) %>%
```

Comment:
nit: I like using `pmax()` and/or `pmin()` for clamping vectors to a range, e.g.
```r
pmax(-1:3, 0)
#> [1] 0 0 1 2 3
```